emcc.py

```diff
@@ -830,6 +830,9 @@ def get_cflags(user_args):
   if settings.EMSCRIPTEN_TRACING:
     cflags.append('-D__EMSCRIPTEN_TRACING__=1')
 
+  if settings.SHARED_MEMORY:
+    cflags.append('-D__EMSCRIPTEN_SHARED_MEMORY__=1')
```

Comment:
I've been trying to move away from adding clang flags in emscripten .. with the goal of having `emcc` and `clang --target=wasm32-unknown-emscripten` be as close a possible.

clang already defines a macro called `_REENTRANT` when building multi-threaded mode, and we also define `__EMSCRIPTEN__` and `__EMSCRIPTEN_PTHREADS__`  in clang itself:
https://github.com/llvm/llvm-project/blob/139215af8e33079a51c95d17a11c42603961d901/clang/lib/Basic/Targets/OSTargets.h#L964-L966

Can we rely on the existing `_REENTRANT` or `__EMSCRIPTEN_PTHREADS__` macros rather than adding a new one?    That would be my prefered approach.  If you really think we need a new macro can we maybe we can define it in a header file instead (or directly in clang like the others):

```
#if _REENTRANT
#define __EMSCRIPTEN_SHARED_MEMORY__
#endif
```

Reply:
I was unable to implement proper testing support without being able to distinguish in C/C++ files when building with targeting Wasm Workers. 

We could punt it to the users to define their own defines themselves (and then manually specify a testing specific shared memory/wasm workers flag), but given that this is a flag that will affect codegen decisions globally in a codebase, it would feel like missing out on an important thing that people will likely need to be able to detect. Having each codebase evolve to do such detection differently in a custom manner reduces consistency across codebases. It is good to have a canonical way to detect this, that way e.g. tutorials and public codebases become more coherent.

This multithreading flag is a bit special, compared to many other Emscripten -s linker features, which won't typically cause a ripple to user codebases to have to build different C/C++ code for each variant. (Many -s linker features do affect what kind of JS library code users will write, but one can readily `#if` on all of those flags, so things are good there)

I know LLVM defines __EMSCRIPTEN_PTHREADS__, and I actually would want not to have that when targeting just SHARED_MEMORY or WASM_WORKERS, since the name implies that people might be making decisions that pthreads is available. But because LLVM/Clang decided to internally add that flag, and use it as a general "targeting shared memory", I was not able to modify it. There is no other way in LLVM/Clang to target multithreading with atomics etc. than to specify `-pthread`, i.e. Clang currently assumes pthread=shared memory go hand in hand.

Reply:
It does not feel good to tie `_REENTRANT` choices to when building against shared memory, because code like

```
#if defined(_REENTRANT) && !defined(__EMSCRIPTEN_PTHREADS__)
#include <emscripten/wasm_worker.h>

...

emscripten_wasm_worker_malloc(...);

#endif
```

is not really expressively saying what it really wants to say.

Reply:
Indeed, it sounds like just passing `-pthread` to clang might not be what we want for `-sSHARED_MEMORY`.  I wonder if we will want to pass different flags to clang at some point?  (Perhaps ` -mattr=+atomics` is closer to what we want?)

I don't want to block this PR, but I just want to register my slight sadness that we need to more injecting of `-D` flags in the core of `emcc.py` (as opposed in clang).   I agree that it would probably take some work on the clang side to make this work the way you want.

Reply:
> Perhaps -mattr=+atomics is closer to what we want?)

I am not sure what this flag does yet. What we should do is something that enables atomics codegen (instead of turning them to singlethreaded scalar accesses), and then enables the multithreaded wasm memory import and ctor semantics, and enables TLS, but does not make any assumptions that the pthreads API itself would be available. (I am not sure if there are any such assumptions in Clang today, but I do not want to assume otherwise either, since `-pthread` does have pthread in its name)

> I don't want to block this PR, but I just want to register my slight sadness that we need to more injecting of -D flags in the core of emcc.py (as opposed in clang)

I think emcc.py is the right place for these flags that should not affect what LLVM does. I am not sure if LLVM itself should be registering `__EMSCRIPTEN__` and `__EMSCRIPTEN_PTHREADS__`. I'd rather see LLVM registering something like `__wasm__` and `__wasm_shared_memory__`, and then Emscripten registering `__EMSCRIPTEN__` and `__EMSCRIPTEN_PTHREADS__`.

That way people could distinguish in code whether they are using Emscripten to drive LLVM/Clang to build to Wasm, versus if they are using something else (or maybe directly) driving LLVM/Clang building to Wasm.